src/qe/mbp/mbp_arrays_tg.cpp

```diff
@@ -0,0 +1,367 @@
+/*++
+Copyright (c) 2022 Microsoft Corporation
+
+Module Name:
+
+    mbp_arrays_tg.cpp
+
+Abstract:
+
+    Apply rules for model based projection for arrays on a term graph
+
+Author:
+
+    Hari Govind V K (hgvk94) 2023-03-07
+
+Revision History:
+
+--*/
+
+#include "qe/mbp/mbp_arrays_tg.h"
+#include "qe/mbp/mbp_qel_util.h"
+#include "qe/mbp/mbp_arrays.h"
+#include "util/obj_hashtable.h"
+#include "util/obj_pair_hashtable.h"
+
+struct mbp_array_tg::impl {
+    typedef std::pair<expr *, expr *> expr_pair;
+    ast_manager& m;
+    array_util m_array_util;
+    mbp::term_graph& m_tg;
+    //TODO: cache mdl evaluation eventhough we extend m_mdl
+    model& m_mdl;
+
+    //set of variables on which to apply MBP rules
+    obj_hashtable<app> &m_vars_set;
+
+    //variables created in the last iteration of MBP application
+    app_ref_vector m_new_vars;
+
+    expr_sparse_mark &m_seen;
+    obj_pair_hashtable<expr, expr> m_seenp;
+
+    bool m_reduce_all_selects;
+    // apply rules that split on model
+    bool m_use_mdl;
+
+    // m_has_store.is_marked(t) if t has a subterm store(v) where v is a variable to be eliminated
+    ast_mark m_has_stores;
+    // variables required for applying rules
+    vector<expr_ref_vector> indices;
+    expr_ref_vector terms, rdTerms;
+
+
+    bool has_var(expr* t) {
+        return contains_vars(t, m_vars_set, m);
+    }
+
+    bool has_arr_var(expr* t) {
+        return contains_vars(t, m_vars_set, m, m_array_util.get_family_id(), ARRAY_SORT);
+    }
+
+    bool is_var(expr* t) {
+        return is_uninterp_const(t) && has_var(t);
+    }
+
+    bool is_wr_on_rhs(expr* e) {
+        return is_app(e) && is_partial_eq(to_app(e)) && is_wr_on_rhs(to_app(e)->get_arg(0), to_app(e)->get_arg(1));
+    }
+
+    bool is_wr_on_rhs(expr* lhs, expr* rhs) {
+        return (is_arr_write(rhs) && !is_arr_write(lhs));
+    }
+
+    bool is_arr_write(expr* t) {
+        if (!m_array_util.is_store(t)) return false;
+        return has_var(to_app(t));
+    }
+
+    //Assumes that has_store has already been called for subexpressions of e
+    bool has_stores(expr* e) {
+        if (m_has_stores.is_marked(e)) return true;
+        if (!is_app(e)) return false;
+        if (m_array_util.is_store(e) && is_var(to_app(e)->get_arg(0))) {
+            m_has_stores.mark(e, true);
+            return true;
+        }
+        for (auto c : *(to_app(e))) {
+            if (m_has_stores.is_marked(c)) {
+                m_has_stores.mark(e, true);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    bool is_rd_wr(expr* t) {
+        if (!m_array_util.is_select(t)) return false;
+        if (m_reduce_all_selects) return m_array_util.is_store(to_app(t)->get_arg(0));
+        return has_stores(to_app(t)->get_arg(0));
+    }
+
+    bool should_create_peq(expr* e) {
```

Comment:
not convinced about "should" rename to `is_peq` or `is_implicit_peq` (you would make it explicit later)?